File: Operating_Systems/CPU-Scheduling/SJF.py
# Preemptive Shortest Job First (SRTF) Scheduling

def preemptive_sjf(processes):
    # Sort by Arrival Time, then Burst Time
    processes.sort(key=lambda x: x['arrival'])
    
    # Initialize variables
    time = 0
    completed = 0
    n = len(processes)
    ready_queue = []
    results = []

    # Add a "remaining_time" field to track remaining burst time
    for p in processes:
        p['remaining_time'] = p['burst']
    
    while completed < n:
        # Add processes that have arrived to the ready queue
        for p in processes:
            if p['arrival'] <= time and p not in ready_queue and p['remaining_time'] > 0:
                ready_queue.append(p)
        
        # Sort ready queue by remaining time
        ready_queue.sort(key=lambda x: x['remaining_time'])
        
        if ready_queue:
            # Pick the process with the shortest remaining time
            current_process = ready_queue[0]
            current_process['remaining_time'] -= 1
            time += 1
            
            # If the process is complete
            if current_process['remaining_time'] == 0:
                current_process['completion'] = time
                completed += 1
                ready_queue.remove(current_process)
        else:
            # If no process is ready, move time forward
            time += 1

    # Calculate Turnaround Time (TAT) and Waiting Time (WT)
    for p in processes:
        p['tat'] = p['completion'] - p['arrival']
        p['wt'] = p['tat'] - p['burst']
        results.append(p)

    return results


# Example Processes
processes = [
    {'name': 'P1', 'arrival': 0, 'burst': 7},
    {'name': 'P2', 'arrival': 2, 'burst': 4},
    {'name': 'P3', 'arrival': 4, 'burst': 1},
    {'name': 'P4', 'arrival': 5, 'burst': 4}
]

# Run SRTF
results = preemptive_sjf(processes)

# Print Results
print("Process | Arrival | Burst | Completion | Turnaround | Waiting")
for p in results:
    print(f"{p['name']:>7} | {p['arrival']:>7} | {p['burst']:>5} | {p['completion']:>10} | {p['tat']:>10} | {p['wt']:>7}")

File: Operating_Systems/CPU-Scheduling/SJF.js
// Implement the SJF CPU scheduling algorithm
// SJF: Shortest Job First
// This is a non-preemptive algorithm
// The process with the smallest execution time is selected for execution next
// If two processes have the same execution time, the process that arrived first is selected
// The process is executed until it is completed
// The average waiting time and average turnaround time are calculated
// The waiting time is the total time a process spends waiting in the ready queue
// The turnaround time is the total time a process takes to complete execution
// The smaller the average waiting time and average turnaround time, the better the algorithm
// The SJF algorithm is optimal for minimizing the average waiting time and average turnaround time
// The SJF algorithm is suitable for batch systems with known execution times
// The SJF algorithm is not suitable for interactive systems with unknown execution times
// The SJF algorithm is not suitable for real-time systems with strict deadlines
// The SJF algorithm is not suitable for preemptive systems with context switching
// The SJF algorithm is not suitable for systems with high arrival rates
// The SJF algorithm is not suitable for systems with high priority processes
// The SJF algorithm is not suitable for systems with aging processes
// The SJF algorithm is not suitable for systems with I/O-bound processes
// The SJF algorithm is not suitable for systems with CPU-bound processes
// The SJF algorithm is not suitable for systems with starvation processes
// The SJF algorithm is not suitable for systems with aging processes


// Function to implement the SJF CPU scheduling algorithm
function SJF(processes) {
    let totalWaitingTime = 0;
    let totalTurnaroundTime = 0;

    let currentTime = 0;
    let currentProcess = null;

    let readyQueue = [];
    let completedQueue = [];
    processes.sort((a, b) => a.arrivalTime - b.arrivalTime);

    while (processes.length > 0 || readyQueue.length > 0 || currentProcess !== null) {
        while (processes.length > 0 && processes[0].arrivalTime <= currentTime) {
            readyQueue.push(processes.shift());
        }
        readyQueue.sort((a, b) => a.executionTime - b.executionTime);

        if (currentProcess === null && readyQueue.length > 0) {
            currentProcess = readyQueue.shift();
            currentProcess.startTime = currentTime;
        }

        if (currentProcess !== null) {
            currentProcess.executionTime--;
            currentTime++;

            if (currentProcess.executionTime === 0) {
                currentProcess.completionTime = currentTime;
                currentProcess.turnaroundTime = currentProcess.completionTime - currentProcess.arrivalTime;
                currentProcess.waitingTime = currentProcess.turnaroundTime - currentProcess.burstTime;
                totalWaitingTime += currentProcess.waitingTime;
                totalTurnaroundTime += currentProcess.turnaroundTime;
                completedQueue.push(currentProcess);
                currentProcess = null;
            }
        } else {
            currentTime++;
        }
    }
    let averageWaitingTime = totalWaitingTime / completedQueue.length;
    let averageTurnaroundTime = totalTurnaroundTime / completedQueue.length;
    return { completedQueue, averageWaitingTime, averageTurnaroundTime };
}

// Example usage:
let processes = [
    { id: 1, arrivalTime: 0, burstTime: 6 },
    { id: 2, arrivalTime: 1, burstTime: 8 },
    { id: 3, arrivalTime: 2, burstTime: 7 },
    { id: 4, arrivalTime: 3, burstTime: 3 },
];
let result = SJF(processes);
console.log(result.completedQueue);
console.log(`Average Waiting Time: ${result.averageWaitingTime}`);
console.log(`Average Turnaround Time: ${result.averageTurnaroundTime}`);

